### PR TITLE
feat: searchable single select

### PR DESCRIPTION
**Pull request by ninariccimarie**

### CODE REVIEW

**Review by chiahou — COMMENTED** [review]
@ninariccimarie Can we hide the ClearButton when the input box is focused regardless of whether an option has been selected or not? The clear button should be used to clear the selected option instead of letters, since users are most likely to press the delete or backspace key to remove letters during typing. By doing so we can avoid confusion and also to avoid the bug that can be reproduced as follows:
1. Select "Cambodia"
2. Click the input box and type "Ma"
3. Click the clear button
4. Click the input box again
5. Expected result: See the whole list/ Actual result: Only the options contain "Ma" appear

File: src/@next/Select/components/SearchableSelectInput/SearchableSelectInputStyle.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  grid-column: 1 / 1;
+  grid-row: 1 / 1;
+
+  font-family: ${NotoSans}, sans-serif;
+  font-style: normal;
+  font-weight: 400;
+  font-size: 16px;
+  line-height: 150%;
+
+  color: ${Neutral.B18};
```
Comment (chiahou) [inline src/@next/Select/components/SearchableSelectInput/SearchableSelectInputStyle.ts:69]: Can we turn this color to the one for placeholder ${Neutral.B40} when the input box is focused before text is entered? To be more precise:
1. An option is selected, we see the selected value in B18
2. The input box is focused, we see the selected value in B40
3. Users start typing, the selected value is replaced by the letters users type in

**Review by chiahou — APPROVED** [review]
all good. qc passed

File: src/@next/Select/Select.tsx [hunk 2]
```diff
@@ -61,21 +66,53 @@ export const Select = ({
 }: SelectProps) => {
   const [popoverActive, setPopoverActive] = useState(false);
   const [optionListHeight, setOptionListHeight] = useState('');
+  const [menuOptions, setMenuOptions] = useState(options);
+  const [inputValue, setInputValue] = useState(searchableProps?.inputValue);
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx]: same as above

File: src/@next/Select/Select.tsx [hunk 3]
```diff
@@ -61,21 +66,53 @@ export const Select = ({
...
+
+  const [searchableSelectState, setSearchableSelectState] =
+    useState<SearchableSelectState>({
+      showSelected: false,
+      showInput: true,
+      showPlaceholder: true,
+    });
+
+  const updateSearchableSelectState = useCallback(
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx]: any reason for useCallback, same as others

File: src/@next/Select/Select.tsx [hunk 4]
```diff
@@ -111,17 +157,18 @@ export const Select = ({
   }, [listHeight, scrollable]);
 
   const activator = () => {
-    if (searchableProps) {
-      const { inputValue, onInputChange } = searchableProps;
+    if (searchable || searchableProps) {
       return (
         <ActivatorTextInputContext.Provider
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx]: why do we still need providers here when we know what is the activator based on searchableProp?

File: src/@next/Select/Select.tsx [hunk 5]
```diff
@@ -134,47 +181,53 @@ export const Select = ({
...
+            )}
+          </ActivatorWrapper>
+        }
+        onClose={handleClose}
+        autofocusTarget="none"
+        preventFocusOnClose
+        fullWidth
+      >
+        <Popover.Pane height={optionListHeight}>
+          <ActivatorTextInputContext.Provider
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx]: Why do we need text provider here but not select? 🤔

File: src/@next/Select/Select.tsx [hunk 6]
```diff
@@ -43,24 +42,54 @@ export interface SelectProps {
...
-  scrollable,
+  scrollable = false,
   sections,
   selectedValues,
   width,
 }: SelectProps) => {
   const [popoverActive, setPopoverActive] = useState(false);
   const [optionListHeight, setOptionListHeight] = useState('');
+  const [menuOptions, setMenuOptions] = useState(options);
+  const [inputValue, setInputValue] = useState(
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx:67]: will this work if user programatically changed the value?

File: src/@next/Select/Select.tsx [hunk 7]
```diff
@@ -43,24 +42,54 @@ export interface SelectProps {
...
+      showSelected: false,
+      showInput: true,
+      showPlaceholder: true,
+    });
+
+  const updateSearchableSelectState = (newState: SearchableSelectState) => {
+    setSearchableSelectState(newState);
+  };
+
+  const updateInputValue = useCallback(
```
Comment (marthayaputra) [inline src/@next/Select/Select.tsx]: can we confirm if we still need the useCallback? Same for others

File: src/@next/Select/Select.tsx (hunk 6 above)
Comment (ninariccimarie) [reply src/@next/Select/Select.tsx:67]: it won't work but we'll be removing this in a separate PR after this one. to handle the input state internally. currently I just added a TODO comment beside the `searchableProps` prop and we'll help to check and replace if it's already been used in employers.

the current way for searchable multi select to update input value is to also pass `searchableProps?.onInputChange`. we also already have it on the storybook code docs

File: src/@next/Select/Select.tsx (hunk 7 above)
Comment (ninariccimarie) [reply src/@next/Select/Select.tsx]: thanks for the catch. this should be removable as well

**Review by marthayaputra — APPROVED** [review]

**Pull request merged**